[#4552] Increase course ID, course name, and session name max length

samuelfangjw opened the pull request:
Fixes #4552 

**Outline of Solution**

<!-- Tell us how you solved the issue. -->
<!-- If there are things you want the reviewers to focus on, include them here as well. -->
<!-- This portion can be skipped if the fix is trivial. -->

Increased max length as follows:
* course id: 40 -> 64
* course name: 64 -> 80
* session name: 38 -> 64

Made changes to frontend to ensure that text does not overflow their containers. This was done mainly by allowing text to wrap if necessary and increasing the size of text containers. Screenshots of some of the UI below. Two test cases were used, names with words "very long..." and names with letters only "MMMMM", M chosen because it is the widest possible letter. This represents a typical case and the extreme case where the maximum possible length is reached. Both cases have the maximum number of characters. Screenshots below were taken on a 13 inch screen. There is slightly less wrapping on larger screens.

<img width="1285" alt="Screenshot 2021-07-16 at 5 44 52 PM" src="https://user-images.githubusercontent.com/60355570/125961542-bb9f40f4-ec6c-410d-b5bb-7093e1e20cf2.png">

In the screenshot above, I was a little unsure where to put the timezone field that was originally on the right side beside the course id. I've placed it below the course id as this is the the current location when the window size is too small. 

<img width="1259" alt="Screenshot 2021-07-16 at 6 13 51 PM" src="https://user-images.githubusercontent.com/60355570/125975851-410fed66-5f96-4749-9e40-873afe2bef98.png">

For the course name, in the common case the text fits over two lines. However in the extreme case the text may take up 3 lines.

![Screenshot 2021-07-17 at 12 05 33 AM](https://user-images.githubusercontent.com/60355570/125976713-d64934b6-59b2-4321-9a61-8f64f3330cc0.png)
![Screenshot 2021-07-17 at 12 03 14 AM](https://user-images.githubusercontent.com/60355570/125976724-56838a86-5f57-45b8-9356-7e9074646121.png)

For edit course details section, the course name overflows the input in the extreme case. This is similar to current behaviour when screen width is too small.

**Notes**
Couldn't seem to get `text-break` css class to work on `<td>` tags.  I added a `td-text-break` class with the same css properties to workaround this issue.

damithc commented:
Nice work @samuelfangjw I have no issues with the UI changes you have posted.

ccyccyccy commented on src/web/app/pages-student/student-home-page/student-home-page.component.scss:
```diff
@@ -1,3 +1,8 @@
   margin: 0;
 }
+
+.td-text-break {
```
Could we use bootstrap's `text-break` instead?

samuelfangjw replied:
There's a [bug](https://github.com/twbs/bootstrap/issues/30803) with `text-break` in the current version of bootstrap css file used (4.5.0) that causes it to not work with certain tags, `<td>` being one of them.

v 4.5.0
<img width="1233" alt="Screenshot 2021-07-19 at 9 15 53 PM" src="https://user-images.githubusercontent.com/60355570/126165585-cccd0e28-2743-4001-98f0-52f21c9bdff4.png">

v 4.4.1
<img width="931" alt="Screenshot 2021-07-19 at 8 25 10 PM" src="https://user-images.githubusercontent.com/60355570/126165621-ae16b707-4a8b-45d5-a284-f495d3578bbf.png">

I used the `td-text-break` class to work around this. If this is not suitable maybe I could wrap all the necessary tags with a span and apply the `text-break` class to the span instead.

ccyccyccy replied:
I think the approach of using `td-text-break` is fine. Just put it in the global `styles.scss` to avoid repeating it everywhere, and leave a comment there to explain the rationale for this class.

> Actually would it be possible to look into changing the version of bootstrap stylesheet used?

What version do you want to change to? Is this issue fixed in a later version of bootstrap? We could create a separate issue for it since there may be breaking changes in changing the version. For this PR, I think we best stick to the current bootstrap version.

ChooJeremy replied:
Changing it is possible but unlikely within the context of one PR, but if many bugs require a change or we do a major patch we can make the change. 

For now, what about moving .td-text-break to a common scss file (such as `page.component`). This way we can also centralize the class name so we can also file an issue/ add to wiki to remove it when we update bootstrap.

samuelfangjw commented on src/web/app/pages-session/session-submission-page/session-submission-page.component.scss:
```diff
@@ -16,7 +16,7 @@
     text-align: right;
   }
 
-  .col-md-6 {
```
So sorry, the class here is not required anymore. I've replaced this with 'text-md-left' and 'text-md-right' classes as similar looking components in other pages all do it this way. e.g. session-result-page. I've removed two styles from this stylesheet, `.col-md-3` and `.col-md.6` as they are no longer necessary. I hope this doesn't count as an unnecessary change.

samuelfangjw commented:
@ccyccyccy @ChooJeremy Thanks for the review, I've taken your suggestions and moved the class to the global `styles.scss` file. I've renamed it `text-break-class` as I thought it would be neater if I was able to use this on all the tags that require it and not just td tags. I've also added a comment about the bug with link to the issue on bootstrap.

ChooJeremy commented on src/web/styles.scss:
```diff
@@ -283,3 +283,11 @@ h6,
+// text-break class on current version of bootstrap (4.5.0) does not work on some tags.
+// This class can be replaced with text-break once bootstrap version is upgraded.
+// See https://github.com/twbs/bootstrap/issues/30803
+.text-break-class {
```
The word `class` here seems useless, since this is marked as a css class with the dot in front of it. What do you think about just `text-break`?

samuelfangjw replied:
My original intention here was to choose a different name from `text-break` as other classes may use the current broken implementation of `text-break` class and I was a bit afraid that changing the behaviour of `text-break` might lead to regressions. 

However, after giving this some thought, maybe it might be better to override the `text-break` class with the correct implementation so that future contributors who use the `text-break` class get the expected implementation. I could also check the all current uses of `text-break` to ensure that there are no regressions.

I'll take a closer look at this and fix the issues above tonight!

samuelfangjw replied:
I renamed `text-break-class` to `text-break` and replaced the implementation with the current [bootstrap implementation of `text-break`](https://github.com/twbs/bootstrap/blob/88980659712653e9d9a00bf370e52f71babe9683/scss/utilities/_text.scss#L65).

I couldn't seem to replicate the issue on the course pages, the text breaks nicely for me. I've updated the implementation though, so hopefully this fixes it!

<img width="1290" alt="Screenshot 2021-07-24 at 12 53 44 AM" src="https://user-images.githubusercontent.com/60355570/126817352-1338e4ad-fd37-47ab-9fe6-e822a648869a.png">

ChooJeremy reviewed (changes requested):
I did some testing on my end and it seems like the courses page hasn't been handled.
<img width="1439" alt="Screen Shot 2021-07-23 at 12 49 56 AM" src="https://user-images.githubusercontent.com/4904988/126677472-3b1a2339-cf60-4f76-84c2-bd2bc0a5781c.png">

ChooJeremy commented on src/web/styles.scss:
```diff
@@ -283,3 +283,11 @@ h6,
 }
+
+// text-break class on current version of bootstrap (4.5.0) does not work on some tags.
+// This class can be replaced with text-break once bootstrap version is upgraded.
```
Your solution seems to work. Do remember to update the comment as well.

ChooJeremy reviewed (changes requested):
Found further issues:
Within the admin search page, searching for the course:
<img width="774" alt="Screen Shot 2021-07-25 at 11 54 25 PM" src="https://user-images.githubusercontent.com/4904988/126905606-b3bb8070-3c32-4812-9a79-beee04b570a1.png">
On trying to create a new feedback session, copying from previous feedback session. 
<img width="770" alt="Screen Shot 2021-07-26 at 12 00 23 AM" src="https://user-images.githubusercontent.com/4904988/126905607-5975cfc9-57c3-43c3-af9d-b77447890cdf.png">
On trying to edit a feedback session, copying a question from another feedback question:
<img width="774" alt="Screen Shot 2021-07-26 at 12 01 15 AM" src="https://user-images.githubusercontent.com/4904988/126905608-9a14d11d-9bbf-40fc-bfee-b5a5ac7f53d0.png">

There may be more. To be comprehensive, other than just manually checking pages, you can consider searching for (ctrl-shift-F in IntelliJ) cases of `courseId`, `courseName` and `feedbackSessionName`, with a `.html` filter.

samuelfangjw commented:
So sorry, i've taken a more thorough look and believe I have fixed all the text overflow issues.

ChooJeremy reviewed (approved):
LGTM; thanks for the help in searching through which pages to update and ensuring they all support this feature!

rrtheonlyone reviewed (approved):
LGTM 💯 Thanks for the work @samuelfangjw and thorough review @ChooJeremy

wkurniawan07 commented:
@samuelfangjw can you resolve the new test failures?

> M chosen because it is the widest possible letter.

It's W, actually, although the difference is rather minor 😄

samuelfangjw commented:
> @samuelfangjw can you resolve the new test failures?
> 
> > M chosen because it is the widest possible letter.
> 
> It's W, actually, although the difference is rather minor 😄

@wkurniawan07 Test failures have been resolved!
